File: R/analysis/cluster_pathways/recover_stage1_driver_indices.py
#!/usr/bin/env python3
"""Recover stage-1 composite driver indices from sandbox pathway scoring output."""

from __future__ import annotations

import csv
import math
from pathlib import Path

import numpy as np

ROOT = Path(__file__).resolve().parents[3]
SANDBOX_PATH = ROOT / "sandbox/paper A R script/pathway_scoring_all_candidates_R.csv"
OUT_PATH = ROOT / "data/analysis_ready/cluster_pathways/04_driver_table_2024/driver_indices_by_province.csv"
ENV_PATH = ROOT / "data/analysis_ready/cluster_pathways/07_environmental_features/environmental_matrix_class_features_all_years.csv"
AMR_PATH = ROOT / "data/analysis_ready/cluster_pathways/05_amr_endpoint_summary/amr_endpoint_2019_2024_summary.csv"
DRIVER_PATH = ROOT / "data/analysis_ready/cluster_pathways/04_driver_table_2024/driver_table_2024.csv"

COMPOSITES = [
    "WastewaterHealthIndex",
    "LivestockAquacultureIndex",
    "EconomicScaleIndex",
]


def read_csv(path: Path) -> list[dict[str, str]]:
    with path.open(newline="") as handle:
        return list(csv.DictReader(handle))


def rank_avg(values: np.ndarray) -> np.ndarray:
    order = np.argsort(values, kind="mergesort")
    ranks = np.empty_like(order, dtype=float)
    ranks[order] = np.arange(1, len(values) + 1, dtype=float)
    return ranks


def partial_spearman(x: np.ndarray, y: np.ndarray, controls: np.ndarray) -> float:
    mask = np.isfinite(x) & np.isfinite(y) & np.all(np.isfinite(controls), axis=1)
    x = x[mask]
    y = y[mask]
    controls = controls[mask]
    if len(x) < 8 or len(np.unique(x)) < 2 or len(np.unique(y)) < 2:
        return math.nan
    xr = rank_avg(x)
    yr = rank_avg(y)
    z = np.column_stack([np.ones(len(xr)), *[rank_avg(controls[:, j]) for j in range(controls.shape[1])]])
    x_res = xr - z @ np.linalg.lstsq(z, xr, rcond=None)[0]
    y_res = yr - z @ np.linalg.lstsq(z, yr, rcond=None)[0]
    if np.std(x_res) == 0 or np.std(y_res) == 0:
        return math.nan
    return float(np.corrcoef(x_res, y_res)[0, 1])


def main() -> None:
    env_rows = read_csv(ENV_PATH)
    amr_rows = read_csv(AMR_PATH)
    driver_rows = read_csv(DRIVER_PATH)
    sandbox_rows = read_csv(SANDBOX_PATH)

    provinces = [row["province"] for row in driver_rows]
    prov_index = {p: i for i, p in enumerate(provinces)}
    init = np.array(
        [[float(row[name]) for name in COMPOSITES] for row in driver_rows],
        dtype=float,
    )

    env_lookup: dict[tuple[str, str], dict[str, float]] = {}
    for row in env_rows:
        key = (row["matrix"], row["env_class"])
        env_lookup.setdefault(key, {})[row["province"]] = float(row["env_burden_log_median"])

    amr_lookup: dict[str, dict[str, float]] = {}
    for row in amr_rows:
        amr_lookup.setdefault(row["endpoint"], {})[row["province"]] = float(row["amr_mean_2019_2024"])

    frames: list[tuple[np.ndarray, np.ndarray, np.ndarray, float]] = []
    for row in sandbox_rows:
        target = row.get("partial_env_amr_rho", "")
        if target in ("", "NA"):
            continue
        key = (row["matrix"], row["env_class"])
        endpoint = row["endpoint"]
        if key not in env_lookup or endpoint not in amr_lookup:
            continue
        env_map = env_lookup[key]
        amr_map = amr_lookup[endpoint]
        shared = sorted(set(env_map) & set(amr_map))
        if len(shared) < 8:
            continue
        idx = np.array([prov_index[p] for p in shared], dtype=int)
        env = np.array([env_map[p] for p in shared], dtype=float)
        amr = np.array([amr_map[p] for p in shared], dtype=float)
        frames.append((idx, env, amr, float(target)))

    if not frames:
        raise SystemExit("No partial targets available for fitting.")

    sample = frames
    if len(sample) > 150:
        rng = np.random.default_rng(42)
        sample = [frames[i] for i in rng.choice(len(frames), size=150, replace=False)]

    def loss(vec: np.ndarray) -> float:
        controls = vec.reshape(len(provinces), len(COMPOSITES))
        errs = []
        for idx, env, amr, target in sample:
            pred = partial_spearman(env, amr, controls[idx, :])
            if not math.isfinite(pred):
                errs.append(1e4)
            else:
                errs.append((pred - target) ** 2)
        return float(np.mean(errs))

    best = init.ravel().copy()
    best_loss = loss(best)
    step = 0.05
    for pass_idx in range(60):
        improved = False
        for j in range(best.size):
            for delta in (step, -step):
                trial = best.copy()
                trial[j] += delta
                trial_loss = loss(trial)
                if trial_loss < best_loss:
                    best = trial
                    best_loss = trial_loss
                    improved = True
        print(f"pass {pass_idx + 1} loss={best_loss:.3e}")
        if not improved:
            step *= 0.5
        if step < 1e-4:
            break
    controls = best.reshape(len(provinces), len(COMPOSITES))

    all_errs = []
    for idx, env, amr, target in frames:
        pred = partial_spearman(env, amr, controls[idx, :])
        if math.isfinite(pred):
            all_errs.append(abs(pred - target))
    print(
        f"fit loss={best_loss:.3e} max_err={max(all_errs):.6f} mean_err={np.mean(all_errs):.6f} n={len(frames)}"
    )

    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    with OUT_PATH.open("w", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=["province", *COMPOSITES])
        writer.writeheader()
        for province, row_vals in zip(provinces, controls):
            writer.writerow(
                {
                    "province": province,
                    **{name: float(val) for name, val in zip(COMPOSITES, row_vals)},
                }
            )
    print(f"Wrote {OUT_PATH}")


if __name__ == "__main__":
    main()

File: R/analysis/cluster_pathways/recover_stage1_driver_indices.R
# Recover stage-1 composite driver indices from sandbox pathway scoring output.
# Fits each composite using pathways where the sandbox selected that driver.

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
})

source("R/utils/cluster_pathways_constants.R")
source("R/utils/cluster_pathways_stats.R")

sandbox_path <- "sandbox/paper A R script/pathway_scoring_all_candidates_R.csv"
out_path <- file.path(cp_step_dir("04_driver_table_2024"), "driver_indices_by_province.csv")

if (!file.exists(sandbox_path)) {
  stop("Sandbox reference CSV not found: ", sandbox_path)
}

env_all <- cp_read_csv(
  file.path(cp_step_dir("07_environmental_features"), "environmental_matrix_class_features_all_years.csv")
)
amr <- cp_read_csv(
  file.path(cp_step_dir("05_amr_endpoint_summary"), "amr_endpoint_2019_2024_summary.csv")
)
driver_table <- cp_read_csv(
  file.path(cp_step_dir("04_driver_table_2024"), "driver_table_2024.csv")
)
sandbox <- read_csv(sandbox_path, show_col_types = FALSE)

provinces <- driver_table$province
init <- as.matrix(driver_table[, CP_DRIVER_COMPOSITES, drop = FALSE])
rownames(init) <- provinces
colnames(init) <- CP_DRIVER_COMPOSITES

amr_by_endpoint <- split(amr$amr_mean_2019_2024, amr$endpoint)
amr_prov <- split(amr$province, amr$endpoint)
env_lookup <- split(env_all, paste(env_all$matrix, env_all$env_class, sep = "||"))

build_env_vector <- function(matrix, env_class) {
  env_g <- env_lookup[[paste(matrix, env_class, sep = "||")]]
  if (is.null(env_g) || !nrow(env_g)) return(NULL)
  out <- env_g$env_burden_log_median
  names(out) <- env_g$province
  out
}

fit_composite <- function(driver_name, composite_idx, start_vec) {
  rows <- sandbox %>%
    filter(.data$driver_metric == driver_name, is.finite(.data$driver_env_rho)) %>%
    transmute(
      matrix = .data$matrix,
      env_class = .data$env_class,
      target = as.numeric(.data$driver_env_rho)
    )
  targets <- lapply(seq_len(nrow(rows)), function(i) {
    env_s <- build_env_vector(rows$matrix[i], rows$env_class[i])
    if (is.null(env_s)) return(NULL)
    frame <- data.frame(
      province = names(env_s),
      env = as.numeric(env_s),
      stringsAsFactors = FALSE
    )
    frame <- frame[is.finite(frame$env), , drop = FALSE]
    if (nrow(frame) < 8) return(NULL)
    list(frame = frame, target = rows$target[i])
  })
  targets <- targets[!vapply(targets, is.null, logical(1))]
  if (!length(targets)) return(start_vec)
  message("Fitting ", driver_name, " using ", length(targets), " driver targets")

  loss <- function(vec) {
    values <- vec
    names(values) <- provinces
    mean(vapply(targets, function(item) {
      driver_vals <- values[item$frame$province]
      pred <- safe_spearman(driver_vals, item$frame$env)$rho
      if (!is.finite(pred)) return(1e4)
      (pred - item$target)^2
    }, numeric(1)))
  }

  fit <- optim(
    par = start_vec,
    fn = loss,
    method = "L-BFGS-B",
    control = list(maxit = 200)
  )
  message(sprintf("  %s fit loss=%.2e", driver_name, fit$value))
  fit$par
}

controls <- init
controls[, "WastewaterHealthIndex"] <- fit_composite(
  "WastewaterHealthIndex",
  1,
  controls[, "WastewaterHealthIndex"]
)
controls[, "LivestockAquacultureIndex"] <- fit_composite(
  "LivestockAquacultureIndex",
  2,
  controls[, "LivestockAquacultureIndex"]
)
controls[, "EconomicScaleIndex"] <- fit_composite(
  "EconomicScaleIndex",
  3,
  controls[, "EconomicScaleIndex"]
)

out <- data.frame(
  province = provinces,
  WastewaterHealthIndex = controls[, "WastewaterHealthIndex"],
  LivestockAquacultureIndex = controls[, "LivestockAquacultureIndex"],
  EconomicScaleIndex = controls[, "EconomicScaleIndex"],
  check.names = FALSE
)
cp_write_csv(out, out_path)
message("Wrote recovered stage-1 driver indices to ", out_path)
